bls12_381/src/g1.rs

```diff
@@ -793,6 +831,370 @@ impl G1Projective {
+
+pub struct G1Compressed([u8; 48]);
+
+impl fmt::Debug for G1Compressed {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        self.0[..].fmt(f)
+    }
+}
+
+impl Default for G1Compressed {
+    fn default() -> Self {
+        G1Compressed([0; 48])
+    }
+}
+
+impl AsRef<[u8]> for G1Compressed {
+    fn as_ref(&self) -> &[u8] {
+        &self.0
+    }
+}
+
+impl AsMut<[u8]> for G1Compressed {
+    fn as_mut(&mut self) -> &mut [u8] {
+        &mut self.0
+    }
+}
+
+pub struct G1Uncompressed([u8; 96]);
+
+impl fmt::Debug for G1Uncompressed {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        self.0[..].fmt(f)
+    }
+}
+
+impl Default for G1Uncompressed {
+    fn default() -> Self {
+        G1Uncompressed([0; 96])
+    }
+}
+
+impl AsRef<[u8]> for G1Uncompressed {
+    fn as_ref(&self) -> &[u8] {
+        &self.0
+    }
+}
+
+impl AsMut<[u8]> for G1Uncompressed {
+    fn as_mut(&mut self) -> &mut [u8] {
+        &mut self.0
+    }
+}
+
+impl Group for G1Projective {
+    type Scalar = Scalar;
+
+    fn random<R: RngCore + ?Sized>(rng: &mut R) -> Self {
+        G1Subgroup::random(rng).into()
+    }
+
+    fn identity() -> Self {
+        Self::identity()
+    }
+
+    fn generator() -> Self {
+        Self::generator()
+    }
+
+    fn is_identity(&self) -> Choice {
+        self.is_identity()
+    }
+
+    #[must_use]
+    fn double(&self) -> Self {
+        self.double()
+    }
+}
+
+impl Group for G1Subgroup {
+    type Scalar = Scalar;
+
+    fn random<R: RngCore + ?Sized>(rng: &mut R) -> Self {
+        loop {
+            let x = Fp::random(rng);
+            let greatest = rng.next_u32() % 2 != 0;
+
+            // Obtain the corresponding y-coordinate given x as y = sqrt(x^3 + 4)
+            let p = ((x.square() * x) + B).sqrt().map(|y| {
+                G1Affine {
+                    x,
+                    // Switch to the correct y-coordinate if necessary.
+                    y: if bool::from(y.lexicographically_largest()) ^ greatest {
+                        -y
+                    } else {
+                        y
+                    },
+                    infinity: 0.into(),
+                }
+            });
+
+            if p.is_some().into() {
+                let p = p.unwrap().to_curve().mul_by_cofactor();
+
+                if bool::from(!p.is_identity()) {
+                    return p;
+                }
+            }
+        }
+    }
+
+    fn identity() -> Self {
+        G1Subgroup(G1Projective::identity())
+    }
+
+    fn generator() -> Self {
+        G1Subgroup(G1Projective::generator())
+    }
+
+    fn is_identity(&self) -> Choice {
+        self.0.is_identity()
+    }
+
+    #[must_use]
+    fn double(&self) -> Self {
+        G1Subgroup(self.0.double())
+    }
+}
+
+impl WnafGroup for G1Projective {
+    fn recommended_wnaf_for_scalar(_scalar: &Self::Scalar) -> usize {
+        let num_bits = Self::Scalar::NUM_BITS as usize;
```

Comment:
I pulled this as-is from the current implementation, but it doesn't make sense to me that we'd be ignoring the scalar and just going from the maximum scalar bit length, unless there is a constant-time concern (but that concern isn't exposed by this function, so it's hard for implementers to reason about those properties here).